src/D2L.CodeStyle.Analyzers/Helpers/WhitelistHelper.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+
+	internal static class WhitelistHelper {
+
+		public static WhitelistHelper<INamedTypeSymbol> TypeLevel(
+			string whitelistFileName
+		) => new WhitelistHelper<INamedTypeSymbol>(
+			whitelistFileName: whitelistFileName,
+			entryFormatter: s => $"{ s.ToString() }, { s.ContainingAssembly.ToDisplayString( SymbolDisplayFormat.MinimallyQualifiedFormat ) }"
+		);
+
+	}
+
+	internal sealed class WhitelistHelper<T> {
+
+		public WhitelistHelper(
+			string whitelistFileName,
+			Func<T, string> entryFormatter
+		) {
+			WhitelistFileName = whitelistFileName;
+			EntryFormatter = entryFormatter;
+		}
+
+		public string WhitelistFileName { get; }
+		public Func<T, string> EntryFormatter { get; }
+
+		public Func<T, bool> LoadWhitelist(
+			AnalyzerOptions options
+		) {
+			ImmutableHashSet<string> whitelist = LoadWhitelist( options.AdditionalFiles );
+			return IsInWhitelist;
+
+			bool IsInWhitelist( T candidate ) {
+				string entry = EntryFormatter( candidate );
+				bool whitelisted = whitelist.Contains( entry );
+				return whitelisted;
+			}

```

Comment:
The idiomatic way to do this in C# is to have a `static Whitelist.Create-` method which loads the whitelist and returns a `Whitelist` object holding on to the hash set and exposing some method like `IsInWhitelist`. Returning a Func like this is unusual... you've given it a name but it's only a local name which is pretty pointless.

Reply:
The API I'd expect for something like this:

```
public sealed class Whitelist {
  public Whitelist( ImmutableHashSet<INamedTypeSymbol> );

  public static Whitelist CreateFromAdditionalFiles( AdditionalFiles );

  public bool IsInWhitelist( INamedTypeSymbol );
```

Reply:
:+1:

Reply:
(If the string stuff is necessary (e.g. ambiguous parsing for generic types) then I'd make the constructor `private` and still keep `Contains` taking a `INamedTypeSymbol` and doing the lowering to strings so that the analyzers don't need to do that.)